pkg/internal/server/grpc/snapshotter_secret.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package grpc
+
+import (
+	"context"
+
+	snapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v6/apis/volumesnapshot/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	AnnVolumeSnapshotClassIsDefault = "snapshot.storage.kubernetes.io/is-default-class"
+	ParamSnapshotterSecretName      = "csi.storage.k8s.io/snapshotter-secret-name"
+	ParamSnapshotterSecretNamespace = "csi.storage.k8s.io/snapshotter-secret-namespace"
+)
+
+// maybeGetSnapshotterSecret returns the snapshotter secret associated with VolumeSnapshotClass, if any.
+// The VolumeSnapshotClass is fetched by name if specified, or else the default VolumeSnapshotClass for the CSI driver is used.
+func (s *Server) maybeGetSnapshotterSecret(ctx context.Context, volumeSnapshotClass *string, driverName string) (map[string]string, error) {
+	secretNs, secretName, err := s.getSnapshotterSecretName(ctx, volumeSnapshotClass, driverName)
+	if err != nil {
+		return nil, err
+	}
+
+	if secretNs == "" {
+		return nil, nil
+	}
+
+	secret, err := s.getSecret(ctx, secretNs, secretName)
+	if err != nil {
+		return nil, err
+	}
+
+	secretMap := map[string]string{}
+
+	for k, v := range secret.Data {
+		secretMap[k] = string(v)
+	}
+
+	return secretMap, nil
+}
+
+// getSnapshotterSecretName returns the name of the snapshotter secret from a VolumeSnapshotClass.
+// The VolumeSnapshotClass is fetched by name if specified, or else the default VolumeSnapshotClass for the CSI driver is used.
+// On successful return, either both secretNs and secretName are set or neither are set.
+func (s *Server) getSnapshotterSecretName(ctx context.Context, volumeSnapshotClass *string, driverName string) (secretNs string, secretName string, err error) {
+	var vsClass *snapshotv1.VolumeSnapshotClass
+
+	if volumeSnapshotClass != nil {
+		vsClass, err = s.getVolumeSnapshotClass(ctx, *volumeSnapshotClass)
+	} else {
+		vsClass, err = s.getDefaultVolumeSnapshotClassForDriver(ctx, driverName)
```

Comment:
I'm handling this case here but from what I can see in the [external-snapshotter common controller code](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/common-controller/snapshot_controller.go#L751) the case is not handled!